### ISSUE TITLE
Modelica.Math.Matrices.LAPACK.dgesvx crashes

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#1743]
**Reported by peter&#x2e;nilsson on 2 Jul 2015 14:51 UTC**
Modelica.Math.Matrices.LAPACK.dgesvx calls the LAPACK procedure dgesvx with an incorrect EQUED argument (10th in order). The input is supposed to be a writable character but instead a string literal "N" is given.
To remedy a new variable declaration:

String equed[1];

needs to be introduced and the "N" (argument nr 10) in the call to dgesvx replaced by this variable.

/Peter

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1743

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 2 Jul 2015 20:25 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 3 Jul 2015 11:52 UTC**
Replying to [ticket:1743 peter.nilsson]:
> Modelica.Math.Matrices.LAPACK.dgesvx calls the LAPACK procedure dgesvx with an incorrect EQUED argument (10th in order). The input is supposed to be a writable character but instead a string literal "N" is given.
> To remedy a new variable declaration:
> 
> String equed[1];
> 
> needs to be introduced and the "N" (argument nr 10) in the call to dgesvx replaced by this variable.
> 
> /Peter

Fixed in fd35e15e708cc9ed6cec7567f2fef31cf87657fc for maintenance and in 433cab0fdf58ec3b099e32f307983f01f9fe45e4 for trunk.

The reported issue is in fact a bug and argument EQUED is an output argument (if FACT='N' which is the case here). The function dgesvx is used in Modelica.Math.Matrices.continuousRiccati and in Modelica.Math.Matrices.discreteRiccati. These two functions are in turn used in the tests of ModelicaTest.Math.Matrices2. Running Matrices2 without this fix and with this fix produces exactly the same result. Strangely enough, the previous version does not crash (therefore, this issue was not recognized before).

**Comment by modelica-trac-importer** [comment]
**Comment by jsten on 7 Jul 2015 06:41 UTC**
After 433cab0fdf58ec3b099e32f307983f01f9fe45e4 we see segmentation when running `ModelicaTest.Math.MatricesInModel`.
The problem seems to be that `dgesvx` expects `EQUED` to be of length one, which it's not.
Initializing the Modelica variable `equed` to a string of length one solves the problem:
```
String equed = " ";
```
Stacktrace:
```
==24950== Process terminating with default action of signal 11 (SIGSEGV)
==24950==  Bad permissions for mapped region at address 0x6447270
==24950==    at 0x63AF9CA: dgesvx_ (in /home/chria/Tests/MSL/binaries/linux64/MatricesInModel.so)
==24950==    by 0x6322525: func_Modelica_Math_Matrices_LAPACK_dgesvx_def22 (MatricesInModel_funcs.c:14378)
==24950==    by 0x631B5CF: func_Modelica_Math_Matrices_continuousRiccati_def19 (MatricesInModel_funcs.c:14049)
==24950==    by 0x618B7EB: func_ModelicaTest_Math_Matrices2_def6 (MatricesInModel_funcs.c:1174)
==24950==    by 0x615745B: model_ode_initialize (MatricesInModel_equ_init.c:18)
==24950==    by 0x637FE9B: jmi_generic_func (jmi_util.c:2438)
==24950==    by 0x63814E1: jmi_ode_initialize (jmi.c:573)
==24950==    by 0x6384B62: jmi_initialize (jmi_me.c:196)
==24950==    by 0x637A226: fmi1_me_initialize (fmi1_me.c:349)
==24950==    by 0x6157092: MatricesInModel_fmiInitialize (MatricesInModel.c:72)
==24950==    by 0x40C2C1: fmi1_me_simulate (in /home/chria/Downloads/FMUChecker-2.0.2b1-linux64/fmuCheck.linux64)
==24950==    by 0x40B407: fmi1_check (in /home/chria/Downloads/FMUChecker-2.0.2b1-linux64/fmuCheck.linux64)
```

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 7 Jul 2015 07:42 UTC**
Fixed in 4fe40393f625b74eb478d9bf14d9efafd7e36ff2 for maintenance and in adc36c8ec41c1ae4f3e7092429eb9abf92b6040c for trunk.
